op-service/dial/active_l2_provider_test.go

```diff
@@ -0,0 +1,104 @@
+package dial
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/ethereum-optimism/optimism/op-service/testutils"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+)
+
+// TestActiveSequencerFailoverBehavior_RollupProvider tests that the ActiveL2RollupProvider
+// will failover to the next provider if the current one is not active.
+func TestActiveSequencerFailoverBehavior_RollupProviders(t *testing.T) {
+	// Create two mock rollup clients, one of which will declare itself inactive after first check.
+	primarySequencer := testutils.MockRollupClient{}
+	primarySequencer.ExpectSequencerActive(true, nil)
+	primarySequencer.ExpectSequencerActive(false, nil)
+	primarySequencer.ExpectClose()
+	secondarySequencer := testutils.MockRollupClient{}
+	secondarySequencer.ExpectSequencerActive(true, nil)
+
+	mockRollupDialer := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (RollupClientInterface, error) {
+		if url == "primary" {
+			return &primarySequencer, nil
+		} else if url == "secondary" {
+			return &secondarySequencer, nil
+		} else {
+			return nil, fmt.Errorf("unknown test url: %s", url)
+		}
+	}
+
+	endpointProvider, err := NewActiveL2RollupProvider(
+		context.Background(),
+		[]string{"primary", "secondary"},
+		1*time.Microsecond,
+		1*time.Minute,
+		testlog.Logger(t, log.LvlDebug),
+		mockRollupDialer,
+	)
+	require.NoError(t, err)
+	// Check that the first client is used, then the second once the first declares itself inactive.
+	firstSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.True(t, &primarySequencer == firstSequencerUsed) // avoids copying the struct (and its mutex, etc.)
+	secondSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.True(t, &secondarySequencer == secondSequencerUsed)
```

Comment:
I suggest to use [`require.Same`](https://pkg.go.dev/github.com/stretchr/testify/require#Same) here and in all other pointer-equality assertions.